Direct gaia hub storage operations

kantai opened the pull request:
This PR implements:

#309 Pass gaia hub root url as part of generating an authentication response
#310 Move gaia hub bearer token generation code from browser into blockstack.js  
#311 Implement direct-to-hub gaia hub storage client

In general, the way it works, is:

1. it bumps the version on the Auth Request
2. tries to read the hubURL from the Auth Response. if it cannot, it defaults to `hub.blockstack.org`
3. stores a user's hubConfig to localStorage

This PR includes a _new_ interface as well, that people may want to advise on. Now that we have a bumped version for the auth request, we need to actually do semantic version checks on those requests. To do this, I added a `isLaterVersion` function to `src/utils.js`

Testing this doesn't require an updated browser (if you use an old browser, the sign in process in the browser will still do the datastore setup and core session token generation even though the app won't use either)

Improvements this should give us:
1. Faster and more reliable writes/reads (and the errors should be more understandable)
2. URLs which match files (i.e., if you putFile(todos.json), the stored file is a `gaia.blockstack.org/hub/<app-addr>/todos.json`
3. Ability to serve applications which use storage over HTTPS (no longer making requests to localhost means no longer having mixed-content errors).

Once the blockstack-browser updates are live, we'll also be able to use storage from the next Windows native build and webapp builds.

Commit:
add gaiahub connection management functions and direct puts/gets

Commit:
and now it works!

Commit:
lint, plus, try to intelligently set the content-type, and use that to return buffer or text

Commit:
clear gaia config on sign outs + export isLaterVersion

Commit:
delete note on a field name

Commit:
export the GaiaHubConfig class

Commit:
Merge branch 'develop' into feature/direct-gaia-hub

Commit:
getFile throws Error when HTTP request fails (matches old behavior)

larrysalibra reviewed (changes requested):
@kantai this works really great for me especially with the fixes you pushed.

A question: can we remove blockstack-storage as a dependency with this pull request?

My two concerns are:

1. That we're potentially leaking unencrypted private keys of users who use the current browser when an app upgrades to this. (previously discussed in an earlier comment)

2. This doesn't have any test coverage. (unless i'm missing it)

Do you think we can make changes to address both of these and still support our time frame?

brandonparee committed:
Add a check inside of `setLocalGaiaHubConnection` to make sure `hubUrl` is contained in `userData`

This change makes it so that after upgrading to the new `direct-gaia-hub` feature users will not need to
log out and back into a blockstack app to use the new Gaia Hub functionality.

kantai committed:
Merge pull request #317 from brandonparee/feature/direct-gaia-hub-login-fix

Compatibility update to PR #314

Commit:
adding unit tests for gaia hub interactions

Commit:
remove blockstack-storage

Commit:
un-bump the version of the AuthRequest

Commit:
404s on fetching file returns `null`

Commit:
adding myself to the contributors

Commit:
changes to suppport

larrysalibra reviewed (commented):
Looks great! Thanks!
